tests/unit/machine/gp/FITCInferenceMethod_unittest.cc

```diff
@@ -590,5 +655,428 @@ TEST(FITCInferenceMethod,get_marginal_likelihood_derivatives_sparse)
+	float64_t weight1=0.02;
+	float64_t weight2=-0.4;
+	float64_t weight3=0;
+	float64_t weight4=0.01;
+	weights(0,0)=weight1;
+	weights(0,1)=weight2;
+	weights(1,0)=weight3;
+	weights(1,1)=weight4;
+	kernel->set_matrix_weights(weights);
+
+	float64_t mean_weight=2.0;
+	CConstMean* mean=new CConstMean(mean_weight);
+
+	// Gaussian likelihood with sigma = 0.5
+	float64_t sigma=0.5;
+	CGaussianLikelihood* lik=new CGaussianLikelihood(sigma);
+
+	// specify GP regression with FITC inference
+	CFITCInferenceMethod* inf=new CFITCInferenceMethod(kernel, features_train,
+		mean, labels_train, lik, latent_features_train);
+
+	float64_t ind_noise=1e-6*CMath::sq(sigma);
+	inf->set_inducing_noise(ind_noise); 
+
+	float64_t scale=4.0;
+	inf->set_scale(scale);
+
+	// build parameter dictionary
+	CMap<TParameter*, CSGObject*>* parameter_dictionary=new CMap<TParameter*, CSGObject*>();
+	inf->build_gradient_parameter_dictionary(parameter_dictionary);
+
+	// compute derivatives wrt parameters
+	CMap<TParameter*, SGVector<float64_t> >* gradient=
+		inf->get_negative_log_marginal_likelihood_derivatives(parameter_dictionary);
+
+	// get parameters to compute derivatives
+	TParameter* scale_param=inf->m_gradient_parameters->get_parameter("scale");
+	TParameter* sigma_param=lik->m_gradient_parameters->get_parameter("sigma");
+	TParameter* noise_param=inf->m_gradient_parameters->get_parameter("inducing_noise");
+	TParameter* weights_param=kernel->m_gradient_parameters->get_parameter("weights");
+		TParameter* mean_param=mean->m_gradient_parameters->get_parameter("mean");
+
+	float64_t dnlZ_sf2=scale*(gradient->get_element(scale_param))[0];
+	float64_t dnlZ_lik=(gradient->get_element(sigma_param))[0];
+	float64_t dnlZ_noise=(gradient->get_element(noise_param))[0];
+	float64_t dnlZ_mean=(gradient->get_element(mean_param))[0];
+
+	SGVector<float64_t> dnlz_weights_vec=gradient->get_element(weights_param);
+	SGMatrix<float64_t> dnlz_weights(dnlz_weights_vec.vector,t_dim, dim, false);
+	//diagonal elements are in the log domain
+	dnlz_weights(0,0)*=weight1;
+	dnlz_weights(1,1)*=weight4;
+
+	dnlZ_lik+=dnlZ_noise;
+
+	// comparison of partial derivatives of negative log marginal likelihood
+	// with result from GPML 3.5 package:
+	//mean: 0.110630684381327
+	//cov: 4.756295043167128
+	//lik: 0.260126513715885
+	//weights=[
+	//-0.326900947549438  -6.119973491296843
+	//0                    0.002268754179272]
+	//Note that in GPML 3.5, weights is a upper-triangular matrix
+	//diagonal elements of the matrix are in the log domain
+	//while off-diagonal elements are in natrual domain
+
+	EXPECT_NEAR(dnlz_weights(0,0), -0.326900947549438, 1E-10);
+	EXPECT_NEAR(dnlz_weights(0,1), -6.119973491296843, 1E-10);
+	//dnlz_weights(1,0) should be different (non-zero)
+	//since the kernel matrix in the Shogun's implementation does not
+	//have the "upper triangular matrix" constraint
+	EXPECT_NEAR(dnlz_weights(1,1), 0.002268754179272, 1E-10);
+
+	EXPECT_NEAR(dnlZ_lik, 0.260126513715885,1E-10);
```

Comment:
technically, there should be seperate tests for all of those variables.
Unit tests should be as small and atomic as possible